rpc_queue.go

```diff
@@ -0,0 +1,170 @@
+	ErrQueuePushOnClosed = errors.New("push on closed rpc queue")
+)
+
+type priorityQueue struct {
+	normal   []*RPC
+	priority []*RPC
+}
+
+func (q *priorityQueue) Len() int {
+	return len(q.normal) + len(q.priority)
+}
+
+func (q *priorityQueue) NormalPush(rpc *RPC) {
+	q.normal = append(q.normal, rpc)
+}
+
+func (q *priorityQueue) PriorityPush(rpc *RPC) {
+	q.priority = append(q.priority, rpc)
+}
+
+func (q *priorityQueue) Pop() *RPC {
+	var rpc *RPC
+
+	if len(q.priority) > 0 {
+		rpc = q.priority[0]
+		q.priority = q.priority[1:]
+	} else if len(q.normal) > 0 {
+		rpc = q.normal[0]
+		q.normal = q.normal[1:]
+	}
+
+	return rpc
+}
+
+type rpcQueue struct {
+	dataAvailable  *sync.Cond
+	spaceAvailable *sync.Cond
+	// Mutex used to access queue
+	queueMu *sync.Mutex
+	queue   *priorityQueue
+
+	// RWMutex used to access closed
+	closedMu *sync.RWMutex
+	closed   bool
+	maxSize  int
+}
+
+func newRpcQueue(maxSize int) *rpcQueue {
+	queueMu := &sync.Mutex{}
+	return &rpcQueue{
+		dataAvailable:  sync.NewCond(queueMu),
+		spaceAvailable: sync.NewCond(queueMu),
+		queueMu:        queueMu,
+		queue:          &priorityQueue{},
+		closedMu:       &sync.RWMutex{},
+		maxSize:        maxSize,
+	}
+}
+
+func (q *rpcQueue) IsClosed() bool {
+	q.closedMu.RLock()
+	defer q.closedMu.RUnlock()
+	return q.closed
+}
+
+func (q *rpcQueue) Push(rpc *RPC, block bool) error {
+	return q.push(rpc, false, block)
+}
+
+func (q *rpcQueue) UrgentPush(rpc *RPC, block bool) error {
+	return q.push(rpc, true, block)
+}
+
+func (q *rpcQueue) push(rpc *RPC, urgent bool, block bool) error {
+	if q.IsClosed() {
+		panic(ErrQueuePushOnClosed)
+	}
+	q.queueMu.Lock()
+	defer q.queueMu.Unlock()
+
+	for q.queue.Len() == q.maxSize {
+		if block {
+			q.spaceAvailable.Wait()
+			// It can receive a signal because the queue is closed.
+			if q.IsClosed() {
+				panic(ErrQueuePushOnClosed)
+			}
+		} else {
+			return ErrQueueFull
+		}
+	}
+	if urgent {
+		q.queue.PriorityPush(rpc)
+	} else {
+		q.queue.NormalPush(rpc)
+	}
+
+	q.dataAvailable.Signal()
+	return nil
+}
+
+// Note that, when the queue is empty and there are two blocked Pop calls, it
+// doesn't mean that the first Pop will get the item from the next Push. The
+// second Pop will probably get it instead.
+func (q *rpcQueue) Pop(ctx context.Context) (*RPC, error) {
+	if q.IsClosed() {
+		return nil, ErrQueueClosed
+	}
+	q.queueMu.Lock()
+	defer q.queueMu.Unlock()
+
+	finished := make(chan struct{})
+	done := make(chan struct{})
+	go func() {
+		select {
+		case <-finished:
+		case <-ctx.Done():
+			// Wake up all the waiting routines. The only routine that correponds
+			// to this Pop call will return from the function. Note that this can
+			// be expensive, if there are too many waiting routines.
+			q.dataAvailable.Broadcast()
+			done <- struct{}{}
+		}
+	}()
+
+	defer func() {
+		// Tell the other routine that this function is finished.
+		select {
+		case finished <- struct{}{}:
+		default:
+		}
+	}()
```

Comment:
Use `context.AfterFunc`.

```suggestion
	defer context.AfterFunc(ctx, func() { q.dataAvailable.Broadcast() })()
```

Reply:
(assuming I'm not missing something important)

Reply:
> (assuming I'm not missing something important)

`q.dataAvailable.Broadcast()` is still called after the function returns even if the contenxt is not done yet. So I don't think `Afterfunc` is the right solution.

Reply:
> (assuming I'm not missing something important)

If you use `AfterFunc`, `q.dataAvailable.Broadcast()` will be called for sure, right? but it may not be called if `finished` is triggered (i.e. the function returns)

Reply:
> but it may not be called if finished is triggered

It won't be called if the function returns, no. But that's by design, right? That's what the current code does and you only want to broadcast on cancel.

Reply:
> That's what the current code does and you only want to broadcast on cancel.

Yes, it's by design. The point of broadcasting is to wake the `q.dataAvailable.Wait()` call up and let the `Pop` routine return.

If it doesn't wake up, the `Pop` routine will not return when the context is cancelled.

We don't want to wake up, if the function returns normally and the context is not cancelled because the `Pop` routine already return.

Reply:
Ok, that's what I thought. That line of code is short for:

```go
	unregisterAfterFunc := context.AfterFunc(ctx, func() { q.dataAvailable.Broadcast() })
	defer unregisterAfterFunc()
```

That is, the function will be unregistered on return.

Reply:
Alright, that makes sense.